### PR TITLE
[Paddle] Some refactor and optimization on fused attention

### PR DESCRIPTION
**Pull request by Wong4j**
1. Get rng state in c++ custom op and extract seed/offset using a CUDA kernel. (similar to PyTorch)
2. Add two fused CUDA kernel to convert `mask` -> `actual seqlens` -> `cu seqlens`.
3. Add more `fused_attention_backend` check in python code.

### CODE REVIEW

**Comment by Wong4j** [comment]
@jeng1220 Please take a look.

**Comment by Wong4j** [comment]
This PR will have some conflicts with [PR#410](https://github.com/NVIDIA/TransformerEngine/pull/410).
Please merge PR#410 first. Then I will fix the conflicts.

**Comment by Wong4j** [comment]
@jeng1220 
Changes in the latest 2 commits:
- Use CUB to do prefix sum
- Use paddle.incubate.nn.layer.fused_dropout_add

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 1]
```diff
@@ -1031,18 +1044,103 @@ __global__ void UpdateScalesKernel(const float *amax, const float *scale, float
...
+    __shared__ int q_smem[BLOCK_SIZE];
+    __shared__ int kv_smem[BLOCK_SIZE];
+    unsigned int tid = threadIdx.x;
+    unsigned int batch_offset = blockIdx.x * q_seqlen * kv_seqlen;
+
+    // load mask, convert to 1/0, add to shared mem
+    q_smem[tid] = 0;
+    for (unsigned int q_idx = tid * kv_seqlen; q_idx < q_seqlen * kv_seqlen;
+         q_idx += BLOCK_SIZE * kv_seqlen) {
+        q_smem[tid] += (mask[q_idx + batch_offset] ? 0 : 1);
```
Comment (jeng1220) [inline transformer_engine/paddle/csrc/custom_ops.cu:1072]: To reduce memory access:
```diff
-    q_smem[tid] = 0;
+    q = 0
    for (unsigned int q_idx = tid * kv_seqlen; q_idx < q_seqlen * kv_seqlen;
         q_idx += BLOCK_SIZE * kv_seqlen) {
-        q_smem[tid] += (mask[q_idx + batch_offset] ? 0 : 1);
+        q += (mask[q_idx + batch_offset] ? 0 : 1);
    }
+   q_smem[tid] = q;
```

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 2]
```diff
@@ -1031,18 +1044,103 @@ __global__ void UpdateScalesKernel(const float *amax, const float *scale, float
...
+    }
+    if (need_kv) {
+        kv_smem[tid] = 0;
+        for (unsigned int kv_idx = tid; kv_idx < kv_seqlen; kv_idx += BLOCK_SIZE) {
+            kv_smem[tid] += (mask[kv_idx + batch_offset] ? 0 : 1);
```
Comment (jeng1220) [inline transformer_engine/paddle/csrc/custom_ops.cu:1077]: same here.

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 3]
```diff
@@ -1031,18 +1044,103 @@ __global__ void UpdateScalesKernel(const float *amax, const float *scale, float
...
+            if (need_kv) kv_smem[tid] += kv_smem[tid + s];
+        }
+        __syncthreads();
+    }
+
+    // write result for this block to global mem
+    if (tid == 0) {
+        q_actual_seqlen[blockIdx.x + 1] = q_smem[0];
+        if (need_kv) {
+            kv_actual_seqlen[blockIdx.x + 1] = kv_smem[0];
```
Comment (jeng1220) [inline transformer_engine/paddle/csrc/custom_ops.cu:1095]: Use `cub::BlockReduce`

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 4]
```diff
@@ -1031,18 +1044,103 @@ __global__ void UpdateScalesKernel(const float *amax, const float *scale, float
...
+            kv_smem[tid] += (mask[kv_idx + batch_offset] ? 0 : 1);
+        }
+    }
+    __syncthreads();
+
+    // do reduction in shared mem
+    for (unsigned int s = blockDim.x / 2; s > 0; s >>= 1) {
+        if (tid < s) {
+            q_smem[tid] += q_smem[tid + s];
+            if (need_kv) kv_smem[tid] += kv_smem[tid + s];
```
Comment (jeng1220) [inline transformer_engine/paddle/csrc/custom_ops.cu:1086]: Use `cub::BlockReduce`

File: transformer_engine/paddle/csrc/custom_ops.cu (hunk 1 above)
Comment (Wong4j) [reply transformer_engine/paddle/csrc/custom_ops.cu]: Done

File: transformer_engine/paddle/csrc/custom_ops.cu (hunk 4 above)
Comment (Wong4j) [reply transformer_engine/paddle/csrc/custom_ops.cu]: Done

**Comment by jeng1220** [comment]
The PR #412 needs to be merged first.

**Comment by jeng1220** [comment]
@timmoon10 ,

There are some UT failures but they are not related to this PR:
* Build / PyTorch (pull_request) 
* Lint / PyTorch Python (pull_request) 

Please review and merge the code if everything looks good.
Thank you.

**Review by timmoon10 — APPROVED** [review]
LGTM, sorry for taking such a long time.

Be advised that https://github.com/NVIDIA/TransformerEngine/pull/403 is adding a more flexible C API for fused attention, although the existing API is still supported.

**Comment by jeng1220** [comment]
> LGTM, sorry for taking such a long time.
> 
> Be advised that #403 is adding a more flexible C API for fused attention, although the existing API is still supported.

It seems that the new #403 changes the order of output dimensions of fused attention.
Here: https://github.com/NVIDIA/TransformerEngine/pull/403/files#diff-8646f274b30b8b4b8c13804b7df0d5032eb53cd033a4fff11aa91053a14f7206R642

It will take some time to study how to integrate it into TE/Paddle.
@timmoon10 , I hope this PR can be merged first.

**Comment by timmoon10** [comment]
That comment was incorrect, attention has already been using `[s, b, h, 3, d]` tensors. I see that Paddle expects `[b, s, 3, h, d]` tensors, so we've found a bug. The Paddle backend transposes the tensors:
https://github.com/NVIDIA/TransformerEngine/blob/3f64de4a7c9c3841ed2dc185042df55ff83c8051/transformer_engine/paddle/layer/attention.py#L341-L343
Could we do something similar for the TE backend?

**Comment by Wong4j** [comment]
Thank you @timmoon10 . I also noticed TE-PyT use `sbh3d` layout by default. 
TE-Paddle only supports `bs3hd` layout for `backend=='transformer_engine'` currently. As for `backend=='paddle'`, I simply split the fused qkv with shape `bs3hd` into three `bshd` tensor, and transpose them to `bhsd`, and do the dot product attention calculation with paddle native op. We have verified in UT that the calculation results of the two backends are consistent. The current implementation is not good enough, but there should be no bug here.

I will continue to improve this feature to support more qkv_layout in next PR, which will involve some big changes to the code. We can merge this PR first.

**Pull request merged**